src/Razor/test/Microsoft.VisualStudio.Razor.IntegrationTests/InProcess/EditorInProcess_Commands.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Razor.IntegrationTests.InProcess;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+using Xunit;
+
+namespace Microsoft.VisualStudio.Extensibility.Testing
+{
+    internal partial class EditorInProcess
+    {
+        public async Task InvokeGoToDefinitionAsync(CancellationToken cancellationToken)
+        {
+            await ExecuteCommandAsync(WellKnownCommandNames.Edit_GoToDefinition, cancellationToken);
+        }
+
+        public async Task CloseDocumentWindowAsync(CancellationToken cancellationToken)
+        {
+            await ExecuteCommandAsync(WellKnownCommandNames.Window_CloseDocumentWindow, cancellationToken);
+        }
+
+        private async Task ExecuteCommandAsync(string commandName, CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var dte = await GetRequiredGlobalServiceAsync<SDTE, EnvDTE.DTE>(cancellationToken);
```

Comment:
💡 DTE should be avoided where possible. In this case, use `IOleCommandTarget`:

```cs
var dispatcher = await GetRequiredGlobalServiceAsync<SUIHostCommandDispatcher, IOleCommandTarget>(cancellationToken);
ErrorHandler.ThrowOnFailure(dispatcher.Exec(commandGuid, commandId, (uint)OLECMDEXECOPT.OLECMDEXECOPT_DODEFAULT, IntPtr.Zero, IntPtr.Zero));
```

Reply:
> DTE should be avoided where possible

Is there a quick summary as to why?

Reply:
In the best case, it's slow. In the worst case, it's broken/outdated/legacy.

DTE is a legacy automation wrapper around other public APIs. The first step to moving to modern APIs is to unwrap the calls so we can understand the true underlying APIs involved.